crates/red_knot_workspace/src/site_packages.rs

```diff
@@ -100,55 +244,341 @@ fn site_packages_dir_from_sys_prefix(
+        let Some(base_executable_home_path) = base_executable_home_path else {
+            return Err(SitePackagesDiscoveryError::PyvenvCfgParseError(
+                pyvenv_cfg_path,
+                PyvenvCfgParseErrorKind::NoHomeKey,
+            ));
+        };
+        let Some(base_executable_home_path) =
+            PythonHomePath::new(base_executable_home_path, system)
+        else {
+            return Err(SitePackagesDiscoveryError::PyvenvCfgParseError(
+                pyvenv_cfg_path,
+                PyvenvCfgParseErrorKind::HomeValueIsNotADirectory,
+            ));
+        };
+
+        // but the `version`/`version_info` key is not read by the standard library,
+        // and is provided under different keys depending on which virtual-environment creation tool
+        // created the `pyvenv.cfg` file. Lenient parsing is appropriate here:
+        // the file isn't really *invalid* if it doesn't have this key,
+        // or if the value doesn't parse according to our expectations.
+        let minor_version = version_info_string.and_then(|version_string| {
+            let mut version_info_parts = version_string.split('.');
+            if version_info_parts.next()? != "3" {
+                return None;
+            }
+            let minor_version_string = version_info_parts.next()?;
+            PythonMinorVersion::from_version_string(minor_version_string)
+        });
+
+        let metadata = Self {
+            venv_path,
+            base_executable_home_path,
+            include_system_site_packages,
+            minor_version,
+        };
+
+        tracing::trace!("Resolved metadata for virtual environment: {metadata:?}");
+        Ok(metadata)
+    }
+
+    /// Return a list of `site-packages` directories that are available from this virtual environment
+    ///
+    /// See the documentation for `site_packages_dir_from_sys_prefix` for more details.
+    pub fn site_packages_dirs(
+        &self,
+        system: &dyn System,
+    ) -> Result<Vec<SystemPathBuf>, SitePackagesDiscoveryError> {
+        let VirtualEnvironment {
+            venv_path,
+            base_executable_home_path,
+            include_system_site_packages,
+            minor_version,
+        } = self;
+
+        let minor_version = minor_version.as_ref().copied();
+
+        let mut site_packages_dirs = vec![site_packages_dir_from_sys_prefix(
+            venv_path,
+            minor_version,
+            system,
+        )?];
+
+        if *include_system_site_packages {
+            let system_sys_prefix =
+                SysPrefixPath::from_executable_home_path(base_executable_home_path, system);
+
+            // If we fail to resolve the `sys.prefix` path from the base executable home path,
+            // or if we fail to resolve the `site-packages` from the `sys.prefix` path,
+            // we should probably print a warning but *not* abort type checking
```

Comment:
In the `ruff_linter` crate I'd use the `warn_user!()` macro for this... what should I use here? A `tracing` call?

Reply:
tracing::warn although there's no guarantee that is only shown once if the function is called multiple times. We may also have to return a diagnostic if this is something that runs in a salsa query. See my tracing documentation

Reply:
I think `tracing::warn!` is fine for this kind of thing? I think a settings-related warning like this is somewhat different from a diagnostic. But @MichaReiser might have a different though.